certora/specs/Tokens.spec

```
@@ -0,0 +1,137 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+import "ConsistentState.spec";
+
+using UtilHarness as Util;
+
+methods {
+    function Util.balanceOf(address, address) external returns(uint256) envfree;
+    function Util.safeTransferFrom(address, address, address, uint256) external envfree;
+    function Util.withdrawnAssets(address, MetaMorphoHarness.Id, uint256, uint256) external returns (uint256) envfree;
+
+    function _.transfer(address, uint256) external => DISPATCHER(true);
+    function _.transferFrom(address, address, uint256) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+
+    function _.supply(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, bytes data) external => summarySupply(marketParams, assets, shares, onBehalf, data) expect (uint256, uint256) ALL;
+    function _.withdraw(MetaMorphoHarness.MarketParams marketParams, uint256 assets, uint256 shares, address onBehalf, address receiver) external => summaryWithdraw(marketParams, assets, shares, onBehalf, receiver) expect (uint256, uint256) ALL;
+    function _.idToMarketParams(MetaMorphoHarness.Id id) external => summaryIdToMarketParams(id) expect MetaMorphoHarness.MarketParams ALL;
+
+    function _.expectedSupplyAssets(MetaMorphoHarness.MarketParams, address) external => NONDET;
+    function _.borrowRate(MetaMorphoHarness.MarketParams, MetaMorphoHarness.Market) external => NONDET;
+}
+
+function summaryIdToMarketParams(MetaMorphoHarness.Id id) returns MetaMorphoHarness.MarketParams {
+    MetaMorphoHarness.MarketParams marketParams;
+    uint256 lastUpdated = Morpho.lastUpdate(id);
+
+    // Safe require because markets in the supply/withdraw queue have positive last update (see LastUpdated.spec).
+    require lastUpdated > 0;
+    // Safe require because it is a verified invariant in Morpho Blue.
+    require lastUpdated > 0 => Morpho.libId(marketParams) == id;
```

maybe worth being more explicit on the state, and explain more in the comments?
```suggestion
    require lastUpdated > 0;
    // Safe require because lastUpdated > 0 => Morpho.libId(marketParams) == id is a verified invariant in Morpho Blue.
    require Morpho.libId(marketParams) == id;
```

---

In the ideal case, this would be a `requireInvariant invariantFromMorpho(marketParams)` so we would not have the choice to change the formula inside. This is not possible currently for a technical reason though. But still I would keep it as is to make it closer to a `requireInvariant` which is sound by default (contrary to `require`).

Completely fine to add more comments though, if something else is unclear

---

get it